runtime/fastly/builtins/fetch/request-response.cpp

```diff
@@ -2513,6 +2514,73 @@ bool Response::bodyUsed_get(JSContext *cx, unsigned argc, JS::Value *vp) {
   return true;
 }
 
+bool Response::ip_get(JSContext *cx, unsigned argc, JS::Value *vp) {
+  METHOD_HEADER(0)
+
+  // non-upstream responses always have undefined IP
+  if (!Response::is_upstream(self)) {
+    args.rval().setUndefined();
+    return true;
+  }
+
+  auto handle = response_handle(self);
+  auto res = handle.get_ip();
+  if (auto *err = res.to_err()) {
+    HANDLE_ERROR(cx, *err);
+    return false;
+  }
+
+  auto ret = std::move(res.unwrap());
+  if (!ret.has_value()) {
+    args.rval().setUndefined();
+    return true;
+  }
+  if (ret->len == 4) {
+    char *out = (char *)malloc(INET_ADDRSTRLEN);
+    inet_ntop(AF_INET, ret->ptr.get(), out, 16);
+    JS::RootedString text(
+        cx, JS_NewStringCopyUTF8N(cx, JS::UTF8Chars(out, strnlen(out, INET_ADDRSTRLEN))));
+    if (!text) {
+      return false;
```

Comment:
It will always be valid ASCII at this point so the only really valid failure would be an OOM error during the copy. We could switch it to an assert instead.